indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamAppenderatorConfig.java

```diff
@@ -0,0 +1,282 @@
+  {
+    return tuningConfig.getIndexSpecForIntermediatePersists();
+  }
+
+  @Override
+  public int getNumPersistThreads()
+  {
+    return tuningConfig.getNumPersistThreads();
+  }
+
+  @Override
+  public Integer getMaxRowsPerSegment()
+  {
+    return tuningConfig.getMaxRowsPerSegment();
+  }
+
+  @Override
+  public Long getMaxTotalRows()
+  {
+    return tuningConfig.getMaxTotalRows();
+  }
+
+  @Override
+  public int getMaxColumnsToMerge()
+  {
+    return maxColumnsToMerge;
+  }
+
+  @Override
+  public boolean equals(Object o)
+  {
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+    SeekableStreamAppenderatorConfig that = (SeekableStreamAppenderatorConfig) o;
+    return maxColumnsToMerge == that.maxColumnsToMerge && Objects.equals(tuningConfig, that.tuningConfig);
+  }
+
+  @Override
+  public int hashCode()
+  {
+    return Objects.hash(tuningConfig, maxColumnsToMerge);
+  }
+
+  /**
+   * Calculates the value to use for "maxColumnsToMerge" when it has been set to
+   * {@link SeekableStreamIndexTaskTuningConfig#DEFAULT_MAX_COLUMNS_TO_MERGE}.
+   */
+  static int calculateDefaultMaxColumnsToMerge(
+      final RuntimeInfo runtimeInfo,
+      final DruidProcessingConfig processingConfig,
+      final TuningConfig tuningConfig
+  )
+  {
+    final int maxColumnsToMergeBasedOnHeapMemory = getMaxColumnsToMergeBasedOnHeapMemory(tuningConfig);
+    final int maxColumnsToMergeBasedOnDirectMemory =
+        getMaxColumnsToMergeBasedOnDirectMemory(runtimeInfo, processingConfig);
+
+    return Math.min(maxColumnsToMergeBasedOnHeapMemory, maxColumnsToMergeBasedOnDirectMemory);
+  }
+
+  /**
+   * Calculates the value to use for "maxColumnsToMerge" based purely on heap memory.
+   */
+  static int getMaxColumnsToMergeBasedOnHeapMemory(TuningConfig tuningConfig)
+  {
+    final long heapMemoryForMerging;
+
+    if (tuningConfig.getMaxBytesInMemory() >= 0) {
+      // maxBytesInMemory is active and represents the amount of heap memory used for indexing.
+      // Use an equal amount of heap memory for merging.
+      heapMemoryForMerging = tuningConfig.getMaxBytesInMemoryOrDefault();
+    } else {
+      // maxBytesInMemory was set to unlimited by the user; use what that _would_ be if it was set to automatic.
+      heapMemoryForMerging = tuningConfig.getAppendableIndexSpec().getDefaultMaxBytesInMemory();
+    }
+
+    return (int) Math.min(heapMemoryForMerging / APPENDERATOR_MERGE_ROUGH_HEAP_MEMORY_PER_COLUMN, Integer.MAX_VALUE);
+  }
+
+  /**
+   * Calculates the value to use for "maxColumnsToMerge" based purely on direct memory.
+   */
+  static int getMaxColumnsToMergeBasedOnDirectMemory(
+      final RuntimeInfo runtimeInfo,
+      final DruidProcessingConfig processingConfig
+  )
+  {
+    final long directMemoryForProcessing =
+        (long) processingConfig.intermediateComputeSizeBytes() *
+        (processingConfig.getNumThreads() + processingConfig.getNumMergeBuffers());
```

Comment:
Could you please explain why we would need to add the numMergeBuffers here? The numMergeBuffers would be used by processing threads only no ?

Reply:
Renamed the variable to `directMemoryForQueries` and added this comment:

> // Query processing requires one processing buffer per thread, plus all the merge buffers.

Hopefully that is more clear. The idea is this is meant to track the size of both the processing pool and merge pool added together.